storage/fcds/fcds.go

```diff
@@ -357,9 +363,45 @@ func (s *Store) shardHasFreeOffsets(shard uint8) (has bool) {
 	return has
 }
 
-// getShard returns a shard number for the chunk address.
-func getShard(addr chunk.Address) (shard uint8) {
-	return addr[len(addr)-1] % shardCount
+// NextShard gets the next shard to write to.
+// uses weighted probability to choose the next shard.
+func (s *Store) NextShard() (shard uint8) {
+	// warning: if multiple writers call this at the same time we might get the same shard again and again
+	// because the free slot value has not been decremented yet(!)
+
+	slots := s.meta.ShardSlots()
+	return probabilisticNextShard(slots)
+}
+
+// probabilisticNextShard returns a next shard to write to
+// using a weighted probability
+func probabilisticNextShard(slots []ShardSlot) (shard uint8) {
+	var sum, movingSum int64
+
+	intervalString := ""
+	for _, v := range slots {
+
+		// we need to consider the edge case where no free slots are available
+		// we still need to potentially insert 1 chunk and so if all shards have
+		// no empty offsets - they all must be considered equally as having at least
+		// one empty slot
+		intervalString += fmt.Sprintf("[%d %d) ", sum, sum+v.Slots+1)
+		sum += v.Slots + 1
+	}
+
+	// do some magic
+	magic := int64(rand.Intn(int(sum)))
+	intervalString = fmt.Sprintf("magic %d, intervals ", magic) + intervalString
+	fmt.Println(intervalString)
```

Comment:
Remove fmt.Println

Reply:
will remove before merge as the unit test is a moving target and this helps to print the intervals to double check everything is in order (and to review :) )